### PR TITLE
Twitch support

### LINKED ISSUE #379: [Feature]: Twitch support
<details>
<summary>Checklist</summary>

- [X] This is a suggestion for LiveTL, not Hyperchat. I will go to [the Hyperchat repo](https://github.com/LiveTL/HyperChat) to suggest a feature for Hyperchat.
- [X] This is a suggestion for the LiveTL web extension, not LiveTL iOS or LiveTL Android. I will go to the [iOS](https://github.com/LiveTL/ios) or [Android](https://github.com/LiveTL/android) repo to suggest features for those platforms.
- [X] I have checked for a duplicate or similar feature request/issue. I made sure to check closed issues as well.
- [X] I am not reporting a bug. If I am, I will use the bug report issue template.
- [X] By submitting this issue, you agree to follow our [Code of Conduct](https://github.com/LiveTL/LiveTL/blob/develop/CODE_OF_CONDUCT.md)

</details>

### Is your feature request related to a problem? Please describe.

KSON is starting Twitch.

### Describe the solution you'd like

Make LiveTL work on Twitch by converting the Twitch chat format to YouTube chat format and slapping in our interface.

### Describe alternatives you've considered

Use LiveTL lite and mchad because Taishi who translates KSON uses mchad.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by ChrRubin**
Adds Twitch support, including with Mchad.

## Features
- Uses MutationObserver for parsing message since tmi.js doesn't work for twitch VODs. My implementation works for both live streams and VODs.
- Popout Mode and Embed Mode. I've also added a hide button since not everyone wants to see the blue bar when they're just watching normal English streamers.
- Settings will pop out when using Embed Mode, for both Twitch and YT.
- Useless settings such as Captions and Chat Zoom are hidden when it is opened from Twitch.

## Known issues
- Verified/Mod/Broadcaster detection does not work on Twitch locales other than English, for the same reasons it used to not work on YT.
- Blocked users on Twitch share the same storage object and filter settings with YT, mainly cause I'm lazy and this just works lol.
- Embed Mode does not look good on light mode Twitch, but who uses light mode anyways.

## Screenshots
| Injected buttons | Embed Mode |
| --- | --- |
| image | image |

image

image

### CODE REVIEW

**Comment by ChrRubin** [comment]
Oh I just realized on my screenshot that @ tags don't show up in the output, have to fix that

**Comment by KentoNishi** [comment]
https://user-images.githubusercontent.com/38841491/150293898-ec702bd6-5b46-4e1c-976d-866bc5691ceb.mp4


@ChrRubin I made changes to the injection script to add resizable panes, can you take a look? Other than that I think it's good to go

**Review by KentoNishi — APPROVED** [review]

File: src/ts/content_scripts/twitch-injector.ts [hunk 1]
```diff
@@ -80,37 +82,117 @@ function injectLtlButtons(frameInfo: Chat.FrameInfo): void {
...
+    resizeBar.style.color = 'white';
+    resizeBar.style.overflow = 'visible';
+    resizeBar.style.justifyContent = 'center';
+    resizeBar.style.alignItems = 'center';
+    resizeBar.style.width = '100%';
+    resizeBar.style.fontSize = '25px';
+    resizeBar.style.display = 'flex';
+    const dots = document.createElement('span');
+    dots.innerText = '⋯';
+    dots.style.transform = 'translate(-2px, -1.5px)';
```
Comment (ChrRubin) [inline src/ts/content_scripts/twitch-injector.ts]: Is this needed for something? The non-centerness is bugging me lmao

File: src/ts/content_scripts/twitch-injector.ts [hunk 2]
```diff
@@ -80,37 +82,117 @@ function injectLtlButtons(frameInfo: Chat.FrameInfo): void {
       vertical-align: middle;
       margin: 0px 5px;
     }
+    .ltl-shifted-svg {
+      transform: translateY(-1px);
```
Comment (ChrRubin) [inline src/ts/content_scripts/twitch-injector.ts]: Same thing here. Unless my eyes are bad this makes it look off center

File: src/ts/sources.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import { writable, Writable } from 'svelte/store';
+import { paramsTabId, paramsFrameId } from '../js/constants';
+
+export function twitchSource(): Writable<Ltl.Message | null> {
+  if (paramsTabId == null || paramsFrameId == null) return writable(null);
+
+  const portName = `${Date.now()}${Math.random()}`;
+
+  // MV3 don't support background scripts
+  const port = chrome.tabs.connect(
```
Comment (r2dev2) [inline src/ts/sources.ts]: `chrome.tabs` doesn't seem to be available when launching embed mode which makes embed mode not work on firefox
undefined chrome.tabs

**Review by r2dev2 — CHANGES_REQUESTED** [review]
TLDR: Everything is more or less fine, but embed does not work on firefox (tried both production and dev) because `chrome.tabs` does not exist for some reason.

Can you make it so that the YouTube LiveTL launcher colors are also dark as the dark variant won in a landslide in the poll?

poll: 1 - 15 for light - dark


Also, should we put some default filters against mod bots or not because when I open up hasanabi's chat, I get bot spam.

bot spamming

**Comment by ChrRubin** [comment]
> Can you make it so that the YouTube LiveTL launcher colors are also dark as the dark variant won in a landslide in the poll?

I'll do this in a separate PR, wanted to clean up some of that code anyway.

> Also, should we put some default filters against mod bots or not because when I open up hasanabi's chat, I get bot spam.

There's not really an easy way to do this, bots usually just show up as moderators. AFAIK not even the Twitch API can identify whether a user is a bot. I guess if we REALLY wanted to do this we'd have to keep a local copy of known bot names, or you'd have to do some weird shit recognizing messages as bot messages.

**Comment by ChrRubin** [comment]
90c4389 adds an option to disable Twitch support globally, which will stop the LTL buttons from being injected, in case there's people who never want to use this.

image

**Comment by KentoNishi** [comment]
@ChrRubin so typechecks are failing, idk why some of the port message objects are untyped? it doesn't even compile now
image

**Comment by ChrRubin** [comment]
> @ChrRubin so typechecks are failing, idk why some of the port message objects are untyped? it doesn't even compile now
> image
> 

@KentoNishi did you recurse pull? I updated the submod

**Comment by KentoNishi** [comment]
OHHHHHH

**Review by r2dev2 — APPROVED** [review]
